Dynamic Notification Policy Routes

msvechla opened the pull request:
See #1789 for the related feature proposal.

This implementation is still work in progress for now and should support the feature proposal.

Updates on `GrafanaNotificationPolicyRoutes` are tracked via watches. I first tried to accomplish this via ownerReferences (see commits), but this is not supported cross-namespace.

---

Example of how the sample policies in `hack/kind/resources/default/grafana-notification-policy.yaml` are rendered:

![CleanShot 2024-12-18 at 11 58 53](https://github.com/user-attachments/assets/b08db111-7bb6-416e-a936-8c9b20402b29)

Status updates on the `GrafanaNotificationPolicy`

```sh
❯ kubectl get grafananotificationpolicy grafananotificationpolicy-sample -o jsonpath="{.status}" |jq
```

```json
{
  "conditions": [
    {
      "lastTransitionTime": "2024-12-17T15:52:31Z",
      "message": "Notification Policy was successfully applied to 1 instances",
      "observedGeneration": 9,
      "reason": "ApplySuccessful",
      "status": "True",
      "type": "NotificationPolicySynchronized"
    }
  ],
  "discoveredRoutes": [
    "grafana-crds/dynamic-c (priority: 1)",
    "default/dynamic-d (priority: 2)",
    "default/dynamic-e (priority: nil)"
  ]
}
```

Events emitted for the merged routes:

```sh
❯ kubectl get events

LAST SEEN   TYPE     REASON   OBJECT                                     MESSAGE
20m         Normal   Merged   grafananotificationpolicyroute/dynamic-c   Route merged into NotificationPolicy default/grafananotificationpolicy-sample
3s          Normal   Merged   grafananotificationpolicyroute/dynamic-d   Route merged into NotificationPolicy default/grafananotificationpolicy-sample
3s          Normal   Merged   grafananotificationpolicyroute/dynamic-e   Route merged into NotificationPolicy default/grafananotificationpolicy-sample
```

msvechla commented:
I have updated the draft PR with the latest discussions from #1789:

- `RouteSelector` has been moved to the `Route` object, allowing to dynamically inject routes on all levels now
- Logic to detect reference loops has been implemented
- Tests have been added accordingly
- `Priority` logic has been removed

There are no checks for `RouteSelector` and `Routes` being mutually exclusive so far.
Unfortunately I think we can not implement this validation on the OpenAPI level, as the `Routes` attribute uses `	// +kubebuilder:validation:Schemaless`, which disables all validation.

Another idea would be to solve this by implementing a [ValidationWebhook](https://sdk.operatorframework.io/docs/building-operators/golang/webhook/). @theSuess do you have any thoughts on this?

### Results of updates Samples

Assembled `hack/kind/resources/default/grafana-notification-policy.yaml`:

![CleanShot 2025-01-21 at 11 06 27](https://github.com/user-attachments/assets/538d7978-58f6-44e0-bc59-26197a2cef76)

```sh
❮ kubectl get grafananotificationpolicy grafananotificationpolicy-sample -o jsonpath="{.status}" |jq
```

```json
{
  "conditions": [
    {
      "lastTransitionTime": "2025-01-21T09:48:19Z",
      "message": "Notification Policy was successfully applied to 1 instances",
      "observedGeneration": 1,
      "reason": "ApplySuccessful",
      "status": "True",
      "type": "NotificationPolicySynchronized"
    }
  ],
  "discoveredRoutes": [
    "default/dynamic-e",
    "grafana-crds/dynamic-c",
    "default/dynamic-d"
  ]
}
```

```sh
❯ kubectl get events

LAST SEEN   TYPE     REASON   OBJECT                                     MESSAGE
20m         Normal   Merged   grafananotificationpolicyroute/dynamic-d   Route merged into NotificationPolicy default/grafananotificationpolicy-sample
20m         Normal   Merged   grafananotificationpolicyroute/dynamic-d   Route merged into NotificationPolicy default/grafananotificationpolicy-sample
20m         Normal   Merged   grafananotificationpolicyroute/dynamic-e   Route merged into NotificationPolicy default/grafananotificationpolicy-sample
20m         Normal   Merged   grafananotificationpolicyroute/dynamic-e   Route merged into NotificationPolicy default/grafananotificationpolicy-sample
```

msvechla commented:
I rebased on the latest master branch.

Additionally I looked into the [ValidationWebhook](https://sdk.operatorframework.io/docs/building-operators/golang/webhook/) for ensuring that `routeSelector` and `routes` are mutually exlusive.

While the implementation of the logic is quite straight-forward, setting up webhooks is a bit more complex, especially bringing this to the Helm chart.

I can certainly do this if you think this is the way forward, we can find another way to do the validation, or maybe skip this validation for now.

Any thoughts?

theSuess commented:
Thanks for your work on this. Validation Webhooks are an area we want to stay clear of. What you can do instead is utilize [validation expressions](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2876-crd-validation-expression-language/README.md). We already have some places where we use this for mutual exclusivity (e.g the [AlertRuleGroup type](https://github.com/grafana/grafana-operator/blob/master/api/v1beta1/grafanaalertrulegroup_types.go#L27)

theSuess commented:
Sorry for the deleted comment, missed the part about api level validation rules in your previous note. 

We had some discussions around Validation Webhooks before and came to the conclusion that these are too finicky to get right.

If the validation rules don't work, I'd just go for a status condition in the respective object when both values are set which tells the user that the dynamic matching takes priority and the `routes` field is ignored

theSuess commented:
The reason why I added the schemaless validation is to support recursive objects. If there is a cleaner way, I'm all for it!

msvechla commented:
I updated the PR and implemented the validation by updating the status condition on both `GrafanaNotificationPolicy` and `GrafanaNotificationPolicyRoute`.

```
Status:
  Conditions:
    Last Transition Time:  2025-01-22T10:08:13Z
    Message:               Dynamically matched definitions from routeSelector will take precedence and the routes field is ignored
    Observed Generation:   8
    Reason:                BothRoutesAndRouteSelectorSpecified
    Status:                True
    Type:                  RoutesIgnoredDueToRouteSelector
```

msvechla commented on controllers/notificationpolicy_controller.go:
```diff
@@ -115,6 +121,23 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
+		}
+		assembledNotificationPolicy, mergedRoutes, err = assembleNotificationPolicyRoutes(ctx, r.Client, namespace, assembledNotificationPolicy)
+		r.Log.Info("assembled notification policy routes", "mergedRoutes", mergedRoutes)
+		if err != nil {
```
How should we handle errors here e.g. due to a loop in matched routes? Is it fine as is, by simply returning with an error? Alternatively we could also update the status condition similar to `applyErrors := make(map[string]string)` below and stop reconciling.

Looking at the code below, it looks like the status condition is never updated when errors occurred, not sure if I am misreading the code or this is on purpose:

```go
        if len(applyErrors) > 0 {
		return ctrl.Result{}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
	}

// ...
	condition := buildSynchronizedCondition("Notification Policy", conditionNotificationPolicySynchronized, notificationPolicy.Generation, applyErrors, len(instances))
	meta.SetStatusCondition(&notificationPolicy.Status.Conditions, condition)
```

Looks like this has been adapted here: https://github.com/grafana/grafana-operator/pull/1815

Any thoughts @theSuess ?

msvechla replied:
Maybe I'm missing something, but the errors are never set on the condition, because the controller always returns when there are errors and never reaches the code to set the error condition:

see: https://github.com/grafana/grafana-operator/blob/0d53646ac1952e891d164a85a0d52d23836cdaf1/controllers/notificationpolicy_controller.go#L134

```go
if len(applyErrors) > 0 {
    return ctrl.Result{}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
}
```

msvechla replied:
I made some modifications here, hopefully that resolves the issue: https://github.com/grafana/grafana-operator/pull/1800/commits/f9247636f4f3a487978dfa704c7e3195cef2fe87

Baarsgaard replied:
@msvechla Good catch it looks like I caused a regression preventing bad synchronization results from being added to the status..
I will fix that ASAP everywhere else.
If you [revert these lines](https://github.com/grafana/grafana-operator/pull/1815/files#diff-a3867b98bdb1a7f628669935c99d18431e65e2e26fead663cd59025b47f7f7daL148-R140) it should fix it and you can then undo the changes to the defer and move `applyErrors` and such back to the Grafana instance for loop

msvechla replied:
IMHOP the way it is implemented now on defer works even better here, because there might be multiple places inside the code where we want to stop the reconcile loop immediately and update the status.

If I revert the lines as you suggested, I can not exit the reconcile loop early when I get an error when assembling the notification policy routes, or I would have to duplicate the statement to build the condition.

Or do you have a better solution?

Baarsgaard replied:
The `synchronize` condition is specifically used when talking about API calls to a Grafana instance.
For other errors, the plan is to re-use or create new shared conditions to use.
Or potentially look into publishing events as you have done here 😉.
Hence, putting it right after the apply/synchronize loop makes the most sense.
a simpler `defer` is generally better also.

This also allows you to remove the `applyError[globalApplyError]` which could confuse users that have a Grafana instance named global(Guilty)

As a side note, dissimilar to other projects' use of conditions, Grafana-Operator removes old conditions immediately rather than setting the status `False`.
This avoids multiple conditions where you have to look for the ones that are `True`

theSuess commented on api/v1beta1/grafananotificationpolicy_types.go:
```diff
@@ -126,6 +129,34 @@ func (r *Route) ToModelRoute() *models.Route {
 }
 
+// isMutuallyExclusive checks if a single route satisfies the mutual exclusivity constraint
+func isMutuallyExclusive(r *Route) bool {
```
nitpick: could be a member function `func (r *Route) SelectorMutuallyExclusive` or `SpecValid`

theSuess commented on api/v1beta1/grafananotificationpolicy_types.go:
```diff
@@ -71,6 +71,9 @@ type Route struct {
 	// repeat interval
 	RepeatInterval string `json:"repeat_interval,omitempty"`
 
+	// Selects GrafanaNotificationPolicyRoutes to merge in when specified
```
Even though we can't validate it, we can mention the mutual exclusivity here

theSuess commented on api/v1beta1/grafananotificationpolicyroute_types.go:
```diff
@@ -0,0 +1,58 @@
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Route for alerts to match against
+	Route *Route `json:"route"`
```
I think the API could be even cleaner by having this inlined so the `.spec` field directly contains the route fields

theSuess commented on controllers/grafananotificationpolicyroute_controller.go:
```diff
@@ -0,0 +1,62 @@
+func (r *GrafanaNotificationPolicyRouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	// TODO(user): your logic here
```
Since we don't do anything with the notification policy route, we can get rid of this entire controller

theSuess commented on controllers/notificationpolicy_controller.go:
```diff
@@ -135,12 +157,104 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
+		r.Log.Error(err, "failed to add merged events to routes")
+	}
+
+	if notificationPolicy.Spec.Route.IsRouteSelectorMutuallyExclusive() {
+		meta.RemoveStatusCondition(&notificationPolicy.Status.Conditions, conditionRoutesIgnoredDueToRouteSelector)
+	} else {
+		setRoutesIgnoredDueToRouteSelectorCondition(&notificationPolicy.Status.Conditions, notificationPolicy.Generation)
+	}
```
Is there anything preventing this check from occurring before applying the notification policy? I'd like to avoid half-applied notification policies

theSuess replied:
Ah I see, I expected an error if both are specified but understand what you're going with here. I'm good with this as well!

theSuess commented on controllers/notificationpolicy_controller.go:
```diff
@@ -135,12 +157,104 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
+// returns an assembled GrafanaNotificationPolicy as well as a list of all merged routes.
+// it ensures that there are no reference loops when discovering routes via labelSelectors
+
+func assembleNotificationPolicyRoutes(ctx context.Context, k8sClient client.Client, namespace *string, notificationPolicy *grafanav1beta1.GrafanaNotificationPolicy) (*grafanav1beta1.GrafanaNotificationPolicy, []*v1beta1.GrafanaNotificationPolicyRoute, error) {
```
nitpick: I think this call signature can be cleaned up a bit
```
func (r *GrafanaNotificationPolicyReconciler) assembleNotificationPolicyRoutes(context.Context,*grafanav1beta1.GrafanaNotificationPolicy)
```

either returning the matched routes or only the notification policy itself

theSuess commented on controllers/notificationpolicy_controller.go:
```diff
@@ -222,6 +336,132 @@ func (r *GrafanaNotificationPolicyReconciler) SetupWithManager(mgr ctrl.Manager)
 			return requests
 		})).
+		Watches(&grafanav1beta1.GrafanaNotificationPolicyRoute{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
+			// resync all notification policies that have a routeSelector that matches the routes labels
```
This is pretty neat! Didn't even know this is possible :)

theSuess commented on controllers/notificationpolicy_controller.go:
```diff
@@ -222,6 +336,132 @@ func (r *GrafanaNotificationPolicyReconciler) SetupWithManager(mgr ctrl.Manager)
+
+// getMatchingNotificationPolicyRoutes retrieves all GrafanaNotificationPolicyRoutes for the given labelSelector
+// results will be limited to namespace when specified
+func getMatchingNotificationPolicyRoutes(ctx context.Context, k8sClient client.Client, labelSelector *metav1.LabelSelector, namespace *string) (*v1beta1.GrafanaNotificationPolicyRouteList, error) {
```
Same suggestion regarding call signature. This can be a struct function for the reconciler

theSuess commented on controllers/notificationpolicy_controller.go:
```diff
@@ -222,6 +336,132 @@ func (r *GrafanaNotificationPolicyReconciler) SetupWithManager(mgr ctrl.Manager)
+
+// getRouteSelectors returns a list of all route selectors specified on a notification policy
+// in either the Route.RouteSelector or any of its Routes
+func getRouteSelectors(route *grafanav1beta1.Route) []*metav1.LabelSelector {
```
nitpick: I think `CollectRouteSelectors` is a better name

theSuess commented on controllers/notificationpolicy_controller.go:
```diff
@@ -135,12 +157,104 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
+			route.RouteSelector = nil
+			for i := range routes.Items {
+				matchedRoute := &routes.Items[i]
+				key := fmt.Sprintf("%s/%s", matchedRoute.Namespace, matchedRoute.Name)
```
nitpick: I'd prefer to implement the `NamespacedResource` for routes and use it here

theSuess commented on docs/docs/alerting/notification-policies.md:
```diff
@@ -9,7 +9,79 @@ For a complete explanation on notification policies, see the [upstream Grafana d
+## Dynamic Notification Policy Routes
+
+There might be scenarios where you can not define the entire notification policy in a single place and you have to assemble it from multiple resouces.
+In this case, you can use the `spec.route.routeSelector` in combination with multiple `GrafanaNotificationPolicyRoute` resources.
```
```suggestion
In this case, you can use the `routeSelector` field in combination with multiple `GrafanaNotificationPolicyRoute` resources.
```

theSuess commented on docs/docs/alerting/notification-policies.md:
```diff
@@ -9,7 +9,79 @@ For a complete explanation on notification policies, see the [upstream Grafana d
+There might be scenarios where you can not define the entire notification policy in a single place and you have to assemble it from multiple resouces.
+In this case, you can use the `spec.route.routeSelector` in combination with multiple `GrafanaNotificationPolicyRoute` resources.
+
+The `routeSelector` can be specified in any `route` object, including the list of `spec.route.routes`.
```
```suggestion
Both `GrafanaNotificationPolicy` and `GrafanaNotificationPolicyRoute` objects support the `routeSelector` field.
```

theSuess commented on docs/docs/alerting/notification-policies.md:
```diff
@@ -9,7 +9,79 @@ For a complete explanation on notification policies, see the [upstream Grafana d
+In case `spec.allowCrossNamespaceImport` is enabled, matching routes will be fetched from all namespaces.
+Otherwise only routes from the same namespace as the `GrafanaNotificationPolicy` will be discovered.
+
+All discovered routes will then get set on the `spec.route.routes[]` of the `GrafanaNotificationPolicy`.
```
```suggestion
The discovered routes are then used when applying the notification policy.
```
Something being set could wrongly mean that the actual resource is updated

theSuess commented on docs/docs/alerting/notification-policies.md:
````diff
@@ -9,7 +9,79 @@ For a complete explanation on notification policies, see the [upstream Grafana d
+        - grafana_folder
+        - alertname
+      routes:
+        - receiver: grafana-default-email
+          object_matchers:
+            - - inline
+              - =
+              - first
+            - - team
+              - =
+              - a
+          routes:
+            - receiver: grafana-default-email
+              object_matchers:
+                - - dynamic
+                  - =
+                  - e
+            - receiver: grafana-default-email
+              object_matchers:
+                - - crossNamespace
+                  - =
+                  - "true"
+                - - dynamic
+                  - =
+                  - c
+              routes:
+                - receiver: grafana-default-email
+                  object_matchers:
+                    - - dynamic
+                      - =
+                      - d
+        - receiver: grafana-default-email
+          object_matchers:
+            - - inline
+              - =
+              - second
+            - - team
+              - =
+              - b
+```
````
I think a screenshot would be nicer here to avoid confusion between kubernetes resources and grafana exports

theSuess reviewed (changes requested):
This looks very good! I've tested it locally and it works as expected. I left some minor comments, but nothing that should require large scale refactoring.

Thanks for also taking care of the documentation right away :100:

Baarsgaard commented:
From a user perspective, I would find an error to be more easy to deal with than partially ignoring parts of the spec I set.
If I was not careful, I could end up seeing half the Contact Points, but not realize the other half is missing.
It could also end up with Contact Points that are not finalized/deleted upon adding a `routeSelector` as the route is schemaless and does not allow immutability validation.

From that point, would it make sense to instead set the `InvalidSpec` condition on `NotificationPolicies` or `PolicyRoutes` where both fields are set?
This would result in `NotificationPolicies` that refuse to be updated and `PolicyRoutes` that are omitted from assembly.

Baarsgaard commented on controllers/notificationpolicy_controller.go:
```diff
@@ -222,6 +344,100 @@ func (r *GrafanaNotificationPolicyReconciler) SetupWithManager(mgr ctrl.Manager)
+	}
+
+	for _, route := range routes {
+		r.Recorder.Event(route, corev1.EventTypeNormal, "Merged", fmt.Sprintf("Route merged into NotificationPolicy %s/%s", notificationPolicy.GetNamespace(), notificationPolicy.GetName()))
```
@theSuess  Are events planned as a replacement instead of status updates or an addition to conditions?
Or is there a reason it has not been used yet?

theSuess replied:
No reason other than it didn't occur to us up until now. I can see this being useful for debugging other resources as well.

Baarsgaard commented on controllers/notificationpolicy_controller.go:
```diff
@@ -135,12 +165,104 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
+// it ensures that there are no reference loops when discovering routes via labelSelectors
+
+func assembleNotificationPolicyRoutes(ctx context.Context, k8sClient client.Client, namespace *string, notificationPolicy *grafanav1beta1.GrafanaNotificationPolicy) (*grafanav1beta1.GrafanaNotificationPolicy, []*v1beta1.GrafanaNotificationPolicyRoute, error) {
+	assembledPolicy := notificationPolicy.DeepCopy()
```
Same here, the `DeepCopy` is currently unnecessary.

msvechla replied:
fixed in https://github.com/grafana/grafana-operator/pull/1800/commits/36c1bede6b24d1c353a0c0917eecd6f497095a1e

Baarsgaard reviewed (changes requested):
Looks really good and I learned quite a few things going through this. 😄

msvechla commented on controllers/notificationpolicy_controller.go:
```diff
@@ -115,7 +130,22 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
 
-	applyErrors := make(map[string]string)
+	var mergedRoutes []*v1beta1.GrafanaNotificationPolicyRoute
+	assembledNotificationPolicy := notificationPolicy.DeepCopy()
```
This is currently required, because we run the `notificationPolicy.Spec.Route.IsRouteSelectorMutuallyExclusive()` at the end of the reconcile, to update the status condition when the selectors are not mutually exclusive.

If we adapt the logic to fail reconciliation, we can move this check to the beginning and avoid the `DeepCopy()` indeed, so I can adapt it

msvechla replied:
fixed in https://github.com/grafana/grafana-operator/pull/1800/commits/36c1bede6b24d1c353a0c0917eecd6f497095a1e

msvechla commented:
> From a user perspective, I would find an error to be more easy to deal with than partially ignoring parts of the spec I set. If I was not careful, I could end up seeing half the Contact Points, but not realize the other half is missing.
> 
> From that point, would it make sense to instead set the `InvalidSpec` condition on `NotificationPolicies` or `PolicyRoutes` where both fields are set? This would result in `NotificationPolicies` that refuse to be updated and/or `PolicyRoutes` that are omitted from assembly depending on where both are configured.
> 
> This is not perfect as some routes might still be missing, but I would see it as less confusing overall on the Policy itself.
> 
> It would mean we can return earlier before attempting the apply as well.

@Baarsgaard I adapted this in the latest commit. I now set the spec to be invalid in these cases and stop the reconcile loop: https://github.com/grafana/grafana-operator/pull/1800/commits/36c1bede6b24d1c353a0c0917eecd6f497095a1e

```
Status:
  Conditions:
    Last Transition Time:  2025-01-27T11:24:47Z
    Message:               RouteSelector and Routes are mutually exclusive
    Observed Generation:   10
    Reason:                FieldsMutuallyExclusive
    Status:                True
    Type:                  InvalidSpec
```

msvechla commented:
I hope I caught everything now, if not let me know!

Thanks a lot for your support! I learned a lot and this was a really nice experience!

msvechla commented:
@Baarsgaard I tried to adapt this in the latest commit https://github.com/grafana/grafana-operator/pull/1800/commits/29f70cd8e1ec54df99a96b97c608820df88114de

- reverted changes that moved building synchronized condition to defer
- added a specific error type for loop detected errors
- set condition for specific assembly errors and emit events otherwise

Thanks for your explanation. If the synchronized condition was built specifically for Grafana API errors, then of course it makes sense to adapt this.

Baarsgaard commented on controllers/notificationpolicy_controller.go:
```diff
@@ -99,6 +116,13 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
+	// check if spec is valid
+	if !notificationPolicy.Spec.Route.IsRouteSelectorMutuallyExclusive() {
+		setInvalidSpecMutuallyExclusive(&notificationPolicy.Status.Conditions, notificationPolicy.Generation)
+		return ctrl.Result{}, nil
```
We should return an error for the log in case users miss the status and rely on the log output for monitoring.

Baarsgaard commented on controllers/notificationpolicy_controller.go:
```diff
@@ -35,25 +37,40 @@ import (
+//+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationpolicyroutes/finalizers,verbs=update
 
 func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	r.Log = log.FromContext(ctx).WithName("GrafanaNotificationPolicyReconciler")
```
#1836 was just merged which refactors how logging is propagated across the entire operator.
Would you try and adapt to the same changes?
```suggestion
	log = log.FromContext(ctx).WithName("GrafanaNotificationPolicyReconciler")
	ctc = log.IntoContext(ctx, log)
```

msvechla replied:
I noticed and just did that 🙂

Baarsgaard reviewed (changes requested):
Huge fan of the changes to the Reconcile function, it's a lot simpler now :D

theSuess reviewed (approved):
Thanks again for all the work on this! I've pushed a small cleanup commit and queued this for merging. Looks like the next release will be full of cool new features!

msvechla commented:
@theSuess do you already know when the next release of the operator including this new feature will happen roughly? Thanks a lot for your insights

theSuess commented:
I don't have a specific date yet, but it'll probably be end of february/beginning of march
